ramips: add support for Cudy WR1300

pyckle opened the pull request:
Specifications:
 - SoC: MediaTek MT7621AT
 - RAM: 128 MB (DDR3)
 - Flash: 16 MB (SPI NOR)
 - WiFi: MediaTek MT7603E, MediaTek MT7612E
 - Switch: 1 WAN, 4 LAN (Gigabit)
 - Ports: 1 USB 3.0
 - Buttons: Reset, WPS
 - LEDs: Power, System, Wan, Lan 1-4, WiFi 2.4G, WiFi 5G, WPS, USB
 - Power: DC 12V 1A tip positive

UART Serial:
  115200 baud
  Located on unpopulated 4 pin header near J4:

  J4
  [o] Rx
  [o] Tx
  [o] GND
  [ ] Vcc - Do not connect

Installation:

Download and flash the manufacturer's built OpenWRT image available at
http://www.cudytech.com/openwrt_software_download
Install the new OpenWRT image via luci (System -> Backup/Flash firmware)

Recovery:
 - Loads only signed manufacture firmware due to bootloader RSA verification
 - serve tftp-recovery image as /recovery.bin on 192.168.1.88/24
 - connect to any lan ethernet port
 - power on the device while holding the reset button
 - wait at least 8 seconds before releasing reset button for image to
   download
 - See http://www.cudytech.com/newsinfo/547425.html

Signed-off-by: Andrew Pikler <andrew.pikler@gmail.com>

pyckle committed:
ramips: add support for Cudy WR1300

Specifications:
 - SoC: MediaTek MT7621AT
 - RAM: 128 MB (DDR3)
 - Flash: 16 MB (SPI NOR)
 - WiFi: MediaTek MT7603E, MediaTek MT7612E
 - Switch: 1 WAN, 4 LAN (Gigabit)
 - Ports: 1 USB 3.0
 - Buttons: Reset, WPS
 - LEDs: Power, System, Wan, Lan 1-4, WiFi 2.4G, WiFi 5G, WPS, USB
 - Power: DC 12V 1A tip positive

UART Serial:
  115200 baud
  Located on unpopulated 4 pin header near J4:

  J4
  [o] Rx
  [o] Tx
  [o] GND
  [ ] Vcc - Do not connect

Installation:

Download and flash the manufacturer's built OpenWRT image available at
http://www.cudytech.com/openwrt_software_download
Install the new OpenWRT image via luci (System -> Backup/Flash firmware)
Be sure to NOT keep settings. The force upgrade may need to be checked
due to differences in router naming conventions.

Recovery:
 - Loads only signed manufacture firmware due to bootloader RSA verification
 - serve tftp-recovery image as /recovery.bin on 192.168.1.88/24
 - connect to any lan ethernet port
 - power on the device while holding the reset button
 - wait at least 8 seconds before releasing reset button for image to
   download
 - See http://www.cudytech.com/newsinfo/547425.html

MAC addresses as verified by OEM firmware:

use   address   source
LAN   *:f0      label
WAN   *:f1      label + 1
2g    *:f0      label
5g    *:f2      label + 2

The label MAC address is found in bdinfo 0xde00.

Signed-off-by: Andrew Pikler <andrew.pikler@gmail.com>

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 can be dropped.

pyckle replied:
Removed.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+
+/ {
+	compatible = "cudy,wr1300", "mediatek,mt7621-soc";
+	model = "CUDY WR1300";
```
Don't think all-caps is correct here?

pyckle replied:
Changed to Cudy.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
```
memory node can be dropped.

pyckle replied:
Changed to Cudy.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_sys: status {
```
led_status instead of led_sys?

pyckle replied:
Changed.

pyckle replied:
Actually, the led is labeled on the front as "Sys" rather than status. Perhaps led_sys is actually a better label?

adschm replied:
I don't care if you do it consistently, i.e.
led_sys: sys {
label = sys
...

pyckle replied:
Fixed with a consistent convention sys.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+		compatible = "gpio-leds";
+
+		led_sys: status {
+			label = "wr1300:green:status";
```
model prefix can be dropped, just green:status etc.

pyckle replied:
Dropped.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Can this go faster?

pyckle replied:
This is what was in the dts provided by the manufacturer for OpenWRT (See the for developers zip here: http://www.cudytech.com/filedownload/284029). Is it worth testing what happens if it is clocked faster? If so, could you direct me to documentation on how to do this, or is it simply trial and error to see if images boot or crash?

adschm replied:
Well, 10 MHz is just a value everybody copies without thinking. Actually, the devices can typically go much faster which boosts read speed substantially. Typically, you can look into same spec sheets if available and/or try what's reasonably possible. Apart from that, you may also check other/similar supported devices, of course.

pyckle replied:
Tested at 104 MHz as per the datasheet https://www.winbond.com/resource-files/w25q128fv%20rev.l%2008242015.pdf

The router boots. Is there any specific stability issue to search for/test?

mans0n replied:
The datasheet says 104MHz is not for READ, so you should have used `m25p,fast-read`... But it seems the device does not complain without it, I had also noticed that phenomenon on my devices in the past. I don't know why, I haven't check it with a logic analyzer. Perhaps the datasheet is wrong, or perhaps the hardware actually uses a lower frequency.

Anyway, if it works, then no problem with it, good. But my concern is that the situation might be different on someone else's device. (See #3159) Unless you are the manufacturer, you can't really be sure if the freq works on all devices. And if the flash fails, it's really hard for a common user to fix it as it'll need a serial access. So I believe we'd better take a conservative approach here.

MT7621 supports only a set of freqs, `sys_freq/n` where sys_freq = 220 and n >= 2. If you set `spi-max-frequency` to 104MHz, the actual freq used in hardware is 73.3MHz. So I suggest setting `spi-max-frequency` to 80MHz (with `m25p,fast-read`), or 50MHz to be safe. You can test the flash speed by `time dd if=/dev/mtdblock3 of=/dev/null bs=64k`.

pyckle replied:
Changed to 80 MHZ with the fast read. Output of benchmark is 
`root@OpenWrt:~# time dd if=/dev/mtdblock3 of=/dev/null bs=64k
248+0 records in
248+0 records out
real	0m 3.56s
user	0m 0.00s
sys	0m 3.51s
root@OpenWrt:~# time dd if=/dev/mtdblock3 of=/dev/null bs=64k
248+0 records in
248+0 records out
real	0m 3.57s
user	0m 0.00s
sys	0m 3.53s
root@OpenWrt:~# 
`

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+				read-only;
+			};
+
+			debug: partition@fd0000 {
```
Please only add the DT labels which are needed/used.

pyckle replied:
Removed the unnecessary labels.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+				read-only;
+			};
+
+			partition@50000 {
```
Please sort properly.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+};
+
+&pcie0 {
+	mt76@0,0 {
```
wifi@0,0
Same below.

pyckle replied:
Fixed.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+	mt76@0,0 {
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
+		mtd-mac-address = <&bdinfo 0xde00>;
```
No address in factory 0x4 that is taken automatically?

pyckle replied:
I checked the factory partition with xxd, the factory mac address is not there. It is present in the bdinfo partition.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x8000>;
+		mtd-mac-address = <&bdinfo 0xde00>;
+		mtd-mac-address-increment = <2>;
```
No address in factory 0x8004 that is taken automatically?

pyckle replied:
The factory MAC address is not there.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+		port@4 {
+			status = "okay";
+			label = "wan";
+			mtd-mac-address = <&factory 0xde00>;
```
If wan has the same address as the rest of the port, you don't need to specify it here.

Would you please add a table of OEM MAC addresses and locations to the commit message as discussed here:
https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

pyckle replied:
Removed, and added table of mac addresses to commit.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
```
state_default can be used directly.

pyckle replied:
Fixed.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,198 @@
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "wdt", "rgmii2", "jtag";
```
These names are outdated.

pyckle replied:
Used the aliases to remove redundancy.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
 TARGET_DEVICES += buffalo_wsr-600dhp
 
+define Device/cudy_wr1300
+  DTS := WR1300
```
This is obviously not needed, as you see by the changed name for the dts file compared to where you copied this.

pyckle replied:
Whoops. Removed.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
+define Device/cudy_wr1300
+  DTS := WR1300
+  IMAGE_SIZE := 15872k
+  SUPPORTED_DEVICES += cudy-wr1300 wr1300
```
This is not needed for OpenWrt, but I'm always fascinated how people invent values for this property.
What's the string for "OEM OpenWrt" as discussed in the installation instructions ...

pyckle replied:
Assuming I checked this correctly (see below), they have the board name set to cudy-wr1300. I renamed the device to cudy-wr1300 (changed hyphen to underscore) so that the board type matches exactly. This is consistent with the naming of some of the xiaomi routers, and will make upgrading from their image easier (no force needed), and remove the need for this unnecessary variable.

```
strings WR1300-OpenWRT19.07-0716.bin | grep supported_devices
{  "metadata_version": "1.0", "supported_devices":["cudy-wr1300"], "version": { "dist": "OpenWrt", "version": "19.07.3", "revision": "r11063-85e04e9f46", "target": "ramips/mt7621", "board": "cudy-wr1300" } }
```

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
+  DTS := WR1300
+  IMAGE_SIZE := 15872k
+  SUPPORTED_DEVICES += cudy-wr1300 wr1300
+  DEVICE_TITLE := Cudy WR1300
```
DEVICE_TITLE has been replaced ...

pyckle replied:
Fixed and moved to manufacturer/model.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
+  IMAGE_SIZE := 15872k
+  SUPPORTED_DEVICES += cudy-wr1300 wr1300
+  DEVICE_TITLE := Cudy WR1300
+  DEVICE_PACKAGES :=  kmod-mt76 kmod-usb2 kmod-usb3 kmod-usb-ledtrig-usbport
```
Two spaces instead of one.

pyckle replied:
Fixed.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,190 @@
+
+		usb {
+			label = "green:usb";
+			gpios = <&gpio 14 GPIO_ACTIVE_HIGH>;
```
The mixed use of GPIO_ACTIVE_LOW and GPIO_ACTIVE_HIGH is correct here?

pyckle replied:
The USB and sys leds both work as expected. I will verify the WPS light works as expected and then resolve this.

pyckle replied:
I verified that the leds function correct with the mixed use on the led page, and the backwards one (wps) functions backwards when the dts has it set to GPIO_ACTIVE_HIGH.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,190 @@
+
+&state_default {
+	gpio {
+		group = "wdt", "rgmii2", "jtag";
```
I end up with i2c instead of rgmii2 here.

adschm replied:
And "group" -> "groups"

pyckle replied:
Done.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
+  IMAGE_SIZE := 15872k
+  DEVICE_VENDOR := Cudy
+  DEVICE_MODEL := WR1300
+  DEVICE_PACKAGES := kmod-mt76 kmod-usb2 kmod-usb3 kmod-usb-ledtrig-usbport
```
You need to use a specific package instead of kmod-mt76 ...

pyckle replied:
Added both of the wireless modules explicitly.

pyckle commented:
Hey @adschm, thanks for your time and the lighting fast review. I think I've addressed all of your concerns. Let me know if there are other necessary changes or ways for me to make this cleaner. Thanks!

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -222,6 +222,15 @@ define Device/buffalo_wsr-600dhp
 endef
 TARGET_DEVICES += buffalo_wsr-600dhp
 
+define Device/cudy-wr1300
```
No, please not. Keep the underscore, this is OpenWrt policy.

Please add back the version with the correct SUPPORTED_DEVICES string.
In addition, please also add $(Device/dsa-migration) as found for other devices. This will force the user to use -F during upgrade, and is necessary anyway ...

pyckle replied:
Changed back.

musashino205 commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,189 @@
+};
+
+&pcie0 {
+	wifi@0,0 {
```
missing a ```compatible``` line

pyckle replied:
added.

musashino205 commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,189 @@
+};
+
+&pcie1 {
+	wifi@1,0 {
```
same as above, missing a ```compatible``` line

mans0n replied:
AFAIK you should use `wifi@0,0` here.

mans0n commented:
I would be happy if you sort key/led nodes by node name or gpio pin number.

pyckle commented:
> I would be happy if you sort key/led nodes by node name or gpio pin number.

Done, I believe.

adschm commented on target/linux/ramips/dts/mt7621_cudy_wr1300.dts:
```diff
@@ -0,0 +1,194 @@
+		led-boot = &led_sys;
+		led-failsafe = &led_sys;
+		led-running = &led_sys;
+		led-upgrade = &led_sys;
```
You can add `label-mac-device = &gmac0;` here.

pyckle replied:
Done.

adschm commented:
please add a comment like "Make sure to _not_ keep settings!" after the flashing instructions in the commit message.

pyckle commented:
> please add a comment like "Make sure to _not_ keep settings!" after the flashing instructions in the commit message.

Done. I believe the force box will need to be checked too - the SUPPORTED_DEVICES variable didn't seem to work. Should I mention that too, or should I work to attempt to make this check pass?

adschm commented:
Well, if the SUPPORTED_DEVICES does not work then please remove it.

pyckle commented:
> Well, if the SUPPORTED_DEVICES does not work then please remove it.

Done. I also documented in the commit it may require a force flashing. I think this is OK. If somebody makes a mistake and flashes the wrong firmware, the bootloader's TFTP recovery should be adequate to prevent bricking.

adschm commented:
Thanks, will be merged during the next hour.
